src/main/java/com/icthh/xm/ms/balance/service/BalanceService.java

```diff
@@ -269,6 +277,216 @@ private Optional<Pocket> findPocketForReload(ReloadBalanceRequest reloadRequest,
+        log.info("Operation type to revert: {}", previousOperationType);
+        if (previousOperationType.equals(RELOAD)) {
+            return revertReloadOperation(balance, revertRequest, operationToRevert);
+        } else if (previousOperationType.equals(CHARGING)) {
+            return revertChargingOperation(balance, revertRequest, operationToRevert);
+        } else {
+            log.error("Revert of {} operation type is not supported", operationToRevert.getOperationType());
+            throw new BusinessException("error.revert.operation.type.not.supported", "Revert operation type not supported");
+        }
+    }
+
+    private BalanceChangeEventDto revertReloadOperation(Balance balance,
+                                                        RevertBalanceOperationRequest revertRequest,
+                                                        BalanceChangeEventDto reloadOperationToRevert) {
+        String operationUuid = StringUtils.isNotBlank(revertRequest.getUuid())
+            ? revertRequest.getUuid()
+            : UUID.randomUUID().toString();
+
+        BalanceChangeEvent changeEvent = createBalanceChangeEvent(
+            operationUuid,
+            reloadOperationToRevert.getAmountTotal(),
+            Metadata.of(revertRequest.getMetadata()),
+            now(clock),
+            balance,
+            true,
+            REVERT_RELOAD
+        );
+        changeEvent.setRevertOperationId(revertRequest.getUuidToRevert());
+
+
+        RevertReloadMode revertReloadMode = revertRequest.getRevertReloadMode();
+        log.info("Revert reload mode: {}", revertReloadMode);
+
+        if (revertReloadMode.equals(DEFAULT) || revertReloadMode.equals(AS_MANY_AS_POSSIBLE_FROM_POCKETS)) {
+            revertReloadByMode(revertReloadMode, reloadOperationToRevert.getPocketChangeEvents(), changeEvent);
+        } else if (revertReloadMode.equals(AS_MANY_AS_POSSIBLE_FROM_BALANCE)) {
+            chargingPockets(balance, reloadOperationToRevert.getAmountTotal(), changeEvent, now(clock), true, null);
+        } else {
+            log.error("Could not perform operation, unknown revert reload mode: {}", revertReloadMode);
+            throw new BusinessException("error.revert.reload.mode.unknown", "Unknown revert reload mode");
+        }
+
+        changeEvent = balanceChangeEventRepository.save(changeEvent);
+
+        return balanceChangeEventMapper.toDto(changeEvent);
+    }
+
+    private void revertReloadByMode(RevertReloadMode revertReloadMode,
+                                    List<PocketChangeEventDto> pocketChangeEventDtoList,
+                                    BalanceChangeEvent changeEvent) {
+        List<PocketChangeEvent> pocketChangeEvents = new ArrayList<>();
+
+        for (PocketChangeEventDto pocketChangeEventDto : pocketChangeEventDtoList) {
+            Pocket pocket = pocketRepository.findOneByIdForUpdate(pocketChangeEventDto.getPocketId())
```

Comment:
copy/paste code

just create get method getPocketForUpdate that throw exception

Reply:
Fixed